레스토랑 상세 페이지 구현하기 - Heejae0704

Heejae0704 opened the pull request:
앱 컴포넌트에 대한 라우팅 테스트 코드 작성부터 시작하였습니다.

Heejae0704 committed:
라우팅 기능 구현을 위한 테스트 코드 작성 (RED)

Heejae0704 committed:
About, Home, Restaurants 페이지 라우팅 구현 (GREEN)

Heejae0704 committed:
레스토랑 클릭 시 세부 정보 보여주는 기능 테스트 작성 (RED)

Heejae0704 committed:
reducer 테스트 코드 작성 (RED)

Heejae0704 committed:
reducer, action, api 기능 구현 (GREEN)

- 전체 동작에 대한 일부 테스트 여전히 RED
- 레스토랑 클릭 시 레스토랑 상세 페이지 보여주는 기능을 위한 reducer 등 기능 구현

Heejae0704 committed:
레스토랑 선택 및 레스토랑 세부 정보 표출 구현 (GREEN)

KJoobin commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,17 @@
+const customRender = () => render(<AboutPage />);
+
+describe('AboutPage', () => {
+  context('with path /about', () => {
```
AboutPage 에 router 관련 설정이 없다면 ` context('with path /about', () => {` 를 쓰지 않는게 더 좋을거같아요 :)

Heejae0704 replied:
넵!

KJoobin commented on src/App.jsx:
```diff
@@ -24,9 +27,18 @@ export default function App() {
+      <Routes>
+        <Route exact path="/" element={<HomePage />} />
+        <Route path="/about" element={<AboutPage />} />
+        <Route path="/restaurants">
+          <Route index element={<RestaurantsPage />} />
+          <Route path=":id" element={<RestaurantPage />} />
+        </Route>
```
👍 잘진행해주셨습니다! 공식문서를 통해 찾으신건가요?!

Heejae0704 replied:
네 Switch가 import가 안된다고 해서, 그 김에 공식문서를 찾아봤어요. 진행하다보니 레스토랑 상세 페이지가 id를 받아야 해서, 그럴 때에는 어떻게 하나 더 찾아보고 요런 구조의 예시를 확인할 수 있었습니다.

직접 URL 입력 시에 해당 페이지(http://localhost:8080/restaurant/1)로 찾아 들어가지 못하는 문제도 있었는데, 그건 강의에서 설명해주신 webpack 설정의 `historyAPIFallback` 설정해주는 것에 더해서, index.html에서 `src="main.js"`를 `src="/main.js"`로 수정하는 것으로 해결했습니다. 코드숨 선배님 중에 같은 문제를 겪으신 분이 있었던 거 같아요 ^^ (https://velog.io/@qkr135qkr/netERRABORTED-404-Not-Found-%EC%97%90%EB%9F%AC)

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,34 +1,107 @@
+    const dispatch = jest.fn();
+    useDispatch.mockImplementation(() => dispatch);
 
-  const { queryByText } = render((
-    <App />
-  ));
+    useSelector.mockImplementation((selector) => selector({
+      regions: [
+        { id: 1, name: '서울' },
+      ],
+      categories: [
+        { id: 1, name: '한식' },
+      ],
+      restaurants: [
+        { id: 1, name: '한국식 초밥' },
+      ],
+      selectedRestaurant: { id: 1, name: '한국식 초밥' },
+      restaurant: {
+        id: 6,
+        categoryId: 1,
+        name: '한국식 초밥',
+        address: '서울 강남구',
+        menuItems: [
+          {
+            id: 18,
+            restaurantId: 6,
+            name: '밥',
+          },
+        ],
+      },
+    }));
```
beforeEach 안에 선언을 하신 이유가 있으신가요?

Heejae0704 replied:
굳이 그럴 필요가 없는데 왜 그랬을까요 ^^; clearAllMocks가 mock 함수가 저장하고 있던 호출 횟수 등의 정보를 리셋하는 것이 아니라 mock 자체를 없애는 것이라고 잘못 이해하고 있었던 거 같습니다.

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,34 +1,107 @@
+      expect(queryByText('헤더')).not.toBeNull();
+    });
+
+    it('shows HOME heading', () => {
```
```suggestion
    it('shows "HOME" heading', () => {
```
개인적으로는 특정 문자열이 나오길 기대하는거라면 문자열 이라는걸 보여주는게 오해가 생기지 않는거 같아요 :)

Heejae0704 replied:
넵 수정하였습니다.

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,34 +1,107 @@
+  });
+
+  context('with path /about', () => {
+    it('shows about page description', () => {
```
```suggestion
    it('shows description', () => {
```

it shows description 까지만 써도 괜찮은거같은데 어떠세요?

Heejae0704 replied:
넵

KJoobin commented on src/NotFoundPage.jsx:
```diff
@@ -0,0 +1,7 @@
+export default function NotFoundPage() {
+  return (
+    <>
```
`<>` Fragment 를 쓰신 이유가 있으신가요?

Heejae0704 replied:
처음에 만들어 두었다가 정리를 못한 코드인 거 같습니다.

KJoobin commented on src/NotFoundPage.test.jsx:
```diff
@@ -0,0 +1,17 @@
+      const { queryByText } = customRender();
+
+      expect(queryByText('404 Not Found')).not.toBeNull();
+    });
```
기존에 작성되어있던 테스트를 copy - paste 하시고 설명을 바꾸지 않으신걸까요?

Heejae0704 replied:
네 ㅠ.ㅠ 수정했습니다.

KJoobin commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,47 @@
+export default function RestaurantContainer() {
+  const { id } = useParams();
+  const { name, address, menuItems } = useSelector(get('restaurant'));
+  const { loading } = useSelector((state) => ({
+    loading: state.loading,
+  }));
```
```suggestion
  const { loading } = useSelector(get('loading'));
```
get 함수를 써서 가지고올수 없는 상황인가요?!

Heejae0704 replied:
loading 자체가 object가 아니라 true/false 값이어서.. `const loading = useSelector(get('loading'));`으로 수정했습니다. object destructuring이 익숙해질랑 말랑 하네요. ^^

KJoobin commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,47 @@
+    <div>
+      {loading ? loadingRender : loadedRender}
+    </div>
+  );
```
저는 react component 안에서 component 를 만드는걸 최대한 지양하는 편인데, 희재님의 의견이 궁긍합니다 :) 

더 간단하게 만들수는 없을까요?!

Heejae0704 replied:
컴포넌트를 분리했습니다. (하는 김에 TDD 절차에 따른 커밋도 연습했습니다.) ^^

KJoobin commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,72 @@
+
+  context('with restaurant', () => {
+    it('RestaurantContainer', () => {
+      useSelector.mockImplementation((selector) => selector({
+        restaurant: {
+          id: 6,
+          categoryId: 1,
+          name: '한국식 초밥',
+          address: '서울 강남구',
+          menuItems: [
+            {
+              id: 18,
+              restaurantId: 6,
+              name: '밥',
+            },
+          ],
+        },
+      }));
```
with restaurant 을 위한 설정값인가요?! 아니면 RestaurantContainer 를 위한 설정값인가요?!

Heejae0704 replied:
with restaurant를 위한 설정값입니다. fixture를 이용해서 번잡함을 정리했고요, 테스트 케이스를 restuarant이 선택된 것이 없을 때 (어떤 이유던) restaurant/:id 페이지가 요청된다면 404 not found 페이지를 보여주는 케이스를 추가(without a restaurant)하고 이를 RestaurantContainer에 구현했습니다. 음.. 결과적으로 또 작은 컴포넌트 조각이 RestaurantContainer에 들어가게 되었는데, 이를 좀 더 잘 처리할 수 있는 방법이 있을까요?

Heejae0704 replied:
Restaurant 컴포넌트에서 Guard Clause로 404 페이지를 리턴하는 방식으로 해결했습니다!

KJoobin commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,72 @@
+  });
+
+  context('with restaurant', () => {
+    it('RestaurantContainer', () => {
```
describe -> context -> it 순으로 읽으면 RestaurantContainer with restaurant RestaurantContainer 인데, 뭘 위한 테스트인지 명확하지 않은거 같습니다 :)

Heejae0704 replied:
테스트를 전반적으로 수정하였습니다.

KJoobin commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,72 @@
+    it('renders now loading', () => {
+      useSelector.mockImplementation((selector) => selector({
+        loading: true,
+        restaurant: {
+          id: 6,
+          categoryId: 1,
+          name: '한국식 초밥',
+          address: '서울 강남구',
+          menuItems: [
+            {
+              id: 18,
+              restaurantId: 6,
+              name: '밥',
+            },
+          ],
+        },
```
restaurant 가 겹치는거같은데 fixture 로 관리하는건 어떠세요 ??

Heejae0704 replied:
넵

KJoobin commented on src/RestaurantPage.test.jsx:
```diff
@@ -0,0 +1,74 @@
+    }));
+  });
+
+  context('with a restaurant to show', () => {
+    it('shows detailed info of the restaurant', () => {
```
"App with a restaurant to show, shows detailed info of the restaurant" 를 보면 뭔가 어색한 느낌이 드는데 어떠신가요?!

Heejae0704 replied:
to show가 중복이네요! 수정했습니다.

KJoobin commented on src/RestaurantsContainer.jsx:
```diff
@@ -1,17 +1,31 @@
 import { get } from './utils';
 
 export default function RestaurantsContainer() {
   const restaurants = useSelector(get('restaurants'));
```
object destructuring 을 사용하시는건 어떠세요?

Heejae0704 replied:
restaurants는 array로 받는 부분인데, object destructuring 어떻게 사용할 수 있을까요?

KJoobin replied:
제가 착각했네요!! 잘 진행해주셨습니다!!

KJoobin reviewed (changes requested):
잘 진행해주신거 같아요 :) 어느정도 감을 잡으신거 같아서 다행입니다!

TDD 는 red - green - refactoring 의 사이클을 도는데 green 과 refactoring 이 같이 진행되는거 같아요!! 
red 와 green 도 단위를 더 잘게 쪼개서 1개의 기능에 관해서 red - green -refactoring 후 다음 기능을 TDD 로 진행하시면 좋을거 같아요!!

ex)
1. About page header 테스트 작성 (red)
2. About page header 작성 (green)
3. About page description 테스트 작성 (red)
4. About page description 작성 (green)
5. About page refactoring (refactoring)

Heejae0704 committed:
Loading 페이지 분리 테스트 작성 (RED)

Heejae0704 committed:
Loading 페이지 작성 (GREEN)

Heejae0704 committed:
RestaurantContainer에 Loading 컴포넌트 적용 (REFACTOR)

Heejae0704 committed:
Restaurant 컴포넌트 분리 테스트 작성 (RED)

Heejae0704 committed:
Restaurant 컴포넌트 분리 구현 (GREEN)

Heejae0704 committed:
RestaurantContainer에 Restaurant 컴포넌트 적용 (REFACTOR)

Heejae0704 committed:
코드 리뷰 내용 반영 (REFACTOR)

- Fixture 사용
- 테스트 description 수정

Heejae0704 committed:
불필요한 selectRestaurant 관련 코드 제거 (REFACTOR)

Heejae0704 commented:
질문이 있습니다! 테스트를 작성하고 구현, 리펙터를 반복하는 과정에서.. 기존에 상위 컴포넌트에 있던 테스트가 하위 컴포넌트로 내려가는 경우들이 종종 있는 거 같은데요.. 특히 Redux Container에서 해야하는 테스트와 그 하위 컴포넌트의 테스트는 어떻게 나눠질 수 있나요? 어렴풋이 state 관련된 내용 (useSelector 쓰는 내용이나 useDispatch 쓰는 내용들?)은 Container에서 테스트하고, 실제 렌더링되는 값들은 하위에서 하는 거 같은데, 실제 테스트를 짜다보면 그게 그거 아닌가 싶을 때가 종종 있어서요.. 조언을 부탁드립니다!

Heejae0704 committed:
id가 잘못된 호출 시 404 페이지 렌더링 레스토랑 컴포넌트에 포함

KJoobin commented:
>  특히 Redux Container에서 해야하는 테스트와 그 하위 컴포넌트의 테스트는 어떻게 나눠질 수 있나요? 어렴풋이 state 관련된 내용 (useSelector 쓰는 내용이나 useDispatch 쓰는 내용들?)은 Container에서 테스트하고, 실제 렌더링되는 값들은 하위에서 하는 거 같은데, 실제 테스트를 짜다보면 그게 그거 아닌가 싶을 때가 종종 있어서요.. 조언을 부탁드립니다!

맞습니다. 테스트를 짜다보면 Container 에서 했던 테스트가 거의 동일하게 View 에서도 진행이 되는걸 확인할수 있습니다 ...

희재님께서 말씀해주신것 처럼 테스트의 주체가 다르다고 생각을 해요!! 하위 컴포넌트(=View Component) 에서는 해당 버튼을 눌렀을때 알맞은 함수가 작동하는지, restaurant를 prop 으로 넘겨준다면 화면에 정상적으로 렌더가 되는지? 등등,, Container 에서는 비즈니스 로직이 정상적으로 작동하는지!!

실제로 개발을하거나, 유지보수를 할때 View component 가 깨지면 (보통 View 가 깨지면 Container 도 같이 깨집니다..) View 로직에 문제가 있다는 판단이 들고, Container 만 깨지면 비즈니스 로직에 문제가 있다 라고 인지하는편이에요!

KJoobin commented on src/App.test.jsx:
```diff
@@ -56,27 +44,27 @@ describe('App', () => {
     });
 
-    it('does not show about page description', () => {
+    it('does not show other page\'s description', () => {
```
```suggestion
    it("does not show other page's description", () => {
```
저는 이럴경우 쌍따옴표 를 씁니다..!

Heejae0704 replied:
넵!! 오 이럴 때는 ESLint도 불평을 안하는군요!

KJoobin commented on src/App.test.jsx:
```diff
@@ -9,44 +9,32 @@ import { useDispatch, useSelector } from 'react-redux';
 
 import App from './App';
 
+import regions from '../fixtures/regions';
```
fixture 는 대문자로 작성하시면 어떨까요?
```suggestion
import REGIONS from '../fixtures/regions';
```

Heejae0704 replied:
네 fixture 사용 가능한 테스트에서는 가급적 fixture 사용하도록 수정하였고요, fixture는 대문자로 constant임을 표기하였습니다.

KJoobin commented on src/RestaurantContainer.jsx:
```diff
@@ -2,46 +2,27 @@ import { useEffect } from 'react';
   }, []);
 
-  const loadedRender = (
+  return (
```
guard clauses 로 처리해주시는게 더 깔끔할거같습니다 :)
```suggestion
    if(loading) {
      return (
        <Loading />
      )
    }
    
  return (
    <Restaurant restaurant={restaurant} />
  )
```

Heejae0704 replied:
넵, return JSX 안에서 if문을 사용할 수 없는 것으로 알고 있어서, guard clause로 loading 확인하면 Loading 리턴, 아니면 Restaurant 리턴으로 수정하였습니다.

```
  if (!restaurant.id) return <NotFoundPage />;

  return (
    <>
      <h2>{name}</h2>
      <p>
        주소:
        {address}
      </p>
      <h3>메뉴</h3>
      <ul>
        {menuItems && menuItems.map((menu) => (
          <li key={menu.id}>{menu.name}</li>
        ))}
      </ul>
    </>
  );
```

KJoobin commented on src/RestaurantsContainer.jsx:
```diff
@@ -1,26 +1,19 @@
+            <Link to={path}>
               {restaurant.name}
             </Link>
           </li>
```
저리면 component 를 분리를 했을거같아요!!

Heejae0704 replied:
넵, Restaurants 컴포넌트로 분리했습니다. Restaurant 각 아이템도 컴포넌트로 분리를 하지는 않았습니다. 나중에 Restaurant가 더 복잡해지면 분리해도 될 것 같습니다. (그 때를 대비해서 기존 Restaurant 컴포넌트는 RestaurantDetail로 이름을 변경했습니다.)

KJoobin commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -27,25 +27,4 @@ describe('RestaurantsContainer', () => {
       expect(container).toHaveTextContent('마법사주방');
     });
```
restaurants 이 없을때의 테스트케이스도 만들어서 진행을 해주시면 좋을거같습니다 :)

Heejae0704 replied:
넵

KJoobin reviewed (changes requested):
훨씬 깔끔해진거같아요!!

Heejae0704 committed:
테스트에 fixture 사용하도록 Refactor 및 cleanup

Heejae0704 committed:
Restaurants 컴포넌트 분리 (RED)

Heejae0704 committed:
Restaurant 컴포넌트 구현 (GREEN)

KJoobin commented on src/NotFoundPage.test.jsx:
```diff
@@ -0,0 +1,17 @@
+const customRender = () => render(<NotFoundPage />);
+
+describe('NotFoundPage', () => {
+  context('without defined path', () => {
```
NotFoundPage 만 단독으로 테스트를 할때는 `context('without defined path'
, () => {` 를 쓰는게 더 어색해보이는데 희재님의 의견은 어떠신가요..?

Heejae0704 replied:
describe - context - it 구조를 유지하려다 보니, 너무 당연한 context를 넣게 되네요. 단순한 컴포넌트라면 describe - test 구조로 해도 되겠죠?

KJoobin replied:
describe - it 구조 보다 describe - test 구조가 더 좋으신가요 ??

Heejae0704 replied:
describe - it이 잘 읽히네요 ^^ 수정했습니다.

KJoobin commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,67 @@
+  });
+
+  context('without restaurant', () => {
+    it('does not render', () => {
```
`it('renders "not found"', () => {` 와 같이 구체적으로 작성해주시면 더 좋을거 같아요 :)

Heejae0704 replied:
넵! `it renders invalid route message` 요렇게 수정해보았습니다.

KJoobin commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,23 @@
+import NotFoundPage from './NotFoundPage';
```
view component 에서 Page component 를 부르는건 좋지 않은거 같아요 :) 상위 계층에서 하위계층을 부르는건 자연스럽지만, 하위 계층에서 상위 계층으로 가는건 부자연스러워 보여요!

-- Page Layer
-- Container Layer
-- View Layer

Heejae0704 replied:
NotFoundPage -> InvalidRoute 라고 컴포넌트 명을 변경하고 내용도 "요청하신 정보를 찾을 수 없습니다"로 변경했습니다. View 레이어인 것을 InvalidRouteView라고 명시해야 할까요?

KJoobin commented on src/RestaurantDetail.jsx:
```diff
@@ -0,0 +1,23 @@
+export default function RestaurantDetail({ restaurant }) {
+  const { name, address, menuItems } = restaurant;
+
+  if (!restaurant.id) return <NotFoundPage />;
```
개인적으로는 에러 메세지가 더 친절했으면 좋겠습니다 :)

Heejae0704 replied:
"요청하신 정보를 찾을 수 없습니다"로 변경했습니다.

KJoobin commented on src/Restaurants.jsx:
```diff
@@ -0,0 +1,20 @@
+        const path = `/restaurants/${id}`;
+
+        return (
+          <li key={id || 0}>
```
`id || 0` 을 하신 이유가 있으신가요??

Heejae0704 replied:
레스토랑이 없을 때나 초기 로딩 시에 restaurant이 없으면 테스트에서 `Warning: Each child in a list should have a unique "key" prop.` 에러가 나서 그랬는데요.. 확인을 다시 해보니, 테스트에서만 에러가 나고 실제 앱에서는 해당 에러가 나오지 않았습니다. 테스트 코드를 다시 살펴보니 state의 restaurants에 빈 array가 아니라 `[ {} ] `이렇게 잘못 넣었더라구요. 해당 코드 수정하고 `id || 0` 코드도 수정하였습니다.

KJoobin commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -1,19 +1,48 @@
-    <RestaurantsContainer />
-  ));
+  context('with restaurants', () => {
+    it('RestaurantsContainer', () => {
```
테스트 설명을 수정하지 않으신건가요?!

Heejae0704 replied:
넵.. ㅠ.ㅠ 수정했습니다. 복붙을 여러 번 하다보니 안고치고 놓치는 부분이 생기네요...

KJoobin commented on src/RestaurantsPage.test.jsx:
```diff
@@ -0,0 +1,55 @@
+      }));
+    });
+
+    it('shows regions', () => {
```
render 와 show 의 차이점에 대해서 설명해 주세요 :)

Heejae0704 replied:
차이점이 없습니다. render로 통일했습니다.

KJoobin commented on src/reducer.test.js:
```diff
@@ -108,4 +117,16 @@ describe('reducer', () => {
+
+  describe('setLoading', () => {
+    it('changes loading status', () => {
+      const initialState = {
+        loading: false,
+      };
```
initialState 의 loading 을 true 로 해야지 테스트가 될거 같아요!

Heejae0704 replied:
넵!

Heejae0704 committed:
NotFoundPage 이름 및 내용 변경 테스트 (RED)

- NotFoundPage -> InvalidRoute
- 404 not found -> 요청하신 정보를 찾을 수 없습니다.

Heejae0704 committed:
Invalid Route 컴포넌트 구현 (GREEN)

Heejae0704 committed:
기타 코드 리뷰 수정사항 적용 (REFACTOR)

- RestaurantDetail 랜더 시 "id || 0" 내용 수정
- 테스트 description 수정
- 테스트 로직 수정 (initial loading state)

Heejae0704 committed:
E2E 테스트 이슈 해결

- 잘못된 route 요청 시 "404 Not Found" 문자열 포함

KJoobin commented on src/InvalidRoute.jsx:
```diff
@@ -0,0 +1,8 @@
+export default function InvalidRoute() {
```
네이밍을 NotFound 로 하는건 어떠신가요 :)

Heejae0704 replied:
넵! NotFound가 더 명확한 거 같습니다.

KJoobin reviewed (approved):
잘 진행해주셨습니다 :)

회고 남겨주시면 머지 할게요!!

Heejae0704 committed:
InvalidRoute -> NotFound 컴포넌트 이름 수정

- 및 테스트 구조 일관성 확인 (describe-context-it or describe-it)
